common/beerocks/bwl/common/key_value_parser.cpp

```diff
@@ -0,0 +1,225 @@
+        }
+    }
+}
+
+void KeyValueParser::parse_multiline(std::stringstream &ss_in, std::list<char> delimiter_list,
+                                     parsed_multiline_t &parsed_multiline)
+{
+    if (delimiter_list.empty()) {
+        return;
+    }
+
+    std::string str_storage;
+    while (std::getline(ss_in, str_storage, delimiter_list.front())) {
+        auto delimiter_list_out(delimiter_list);
+
+        // Delete first delimiter
+        delimiter_list_out.erase(delimiter_list_out.begin());
+        std::stringstream ss_out(str_storage);
+        parsed_line_t parsed_line;
+        parse_line(ss_out, delimiter_list_out, parsed_line);
+        parsed_multiline.push_back(parsed_line);
+    }
+}
+
+void KeyValueParser::parse_event(std::string event_str, parsed_line_t &parsed_line)
+{
+    // Eliminate event log level from the begining of the event string : "<3>".
+    size_t idx_start = event_buffer_start_process_idx(event_str);
+
+    // An event string might start with keyless values such as Even Name.
+    // Find where the keyless parameters ends by finding where is the first parameter with key,
+    // assuming it has a preceding " <SomeKeyString>='".
+    auto idx = find_first_of_delimiter_pair(event_str, idx_start, ' ', '=');
+
+    // Put null terminator at the end of our key=val, for string stream construction.
+    if (idx != std::string::npos) {
+        event_str[idx] = '\0';
+    }
+
+    // Parse keyless params part and insert to the 'parsed_line' map object.
+    parse_event_keyless_params(event_str, idx_start, parsed_line);
+
+    // Fill the map with the rest of event data.
+    while (idx != std::string::npos) {
+
+        idx_start = ++idx;
+
+        // Find first '=' to skip on it
+        idx = event_str.find_first_of('=', idx_start);
+
+        // Find the next pair of delimiters index.
+        idx = find_first_of_delimiter_pair(event_str, ++idx, ' ', '=');
+
+        if (idx != std::string::npos) {
+            // Put null terminator at the end of our key=val, for ss_in construction.
+            event_str[idx] = '\0';
+        }
+
+        // Parse key=value
+        std::stringstream ss_in(event_str.c_str() + idx_start);
+        parse_line(ss_in, {'='}, parsed_line);
+    }
+}
+
+size_t KeyValueParser::event_buffer_start_process_idx(const std::string &event_str)
+{
+    // Eliminate event log level from the begining of the event string : "<3>"
+    auto idx_start = event_str.find_first_of(">");
+    if (idx_start != std::string::npos) {
+        idx_start++;
+    } else {
+        LOG(WARNING) << "Empty event! event_string: " << event_str;
+    }
+    return idx_start;
+}
+
+void KeyValueParser::parse_event_keyless_params(const std::string &event_str, size_t idx_start,
+                                                parsed_line_t &parsed_line)
+{
+    // Insert to map known prams without key
+    std::stringstream ss(event_str.c_str() + idx_start);
```

Comment:
This is a bit icky to say the least...

We could actually very easily parse the entire event (including start thing) with a relatively simple istream construct:

```
std::string dummy;
std::getline(in, dummy, ">");
// We could check here that dummy starts with < and is an integer
```

Reply:
I preferred to use `str::string find_first_of()` because the difference between this and what you suggested is a construction  of an additional string (and also copying the string into it).